### PR TITLE
S2 display ingredient

### PR DESCRIPTION
**Pull request by DanielGardiner**

### CODE REVIEW

**Comment by DanielGardiner** [comment]
- added landing page front-end (displays page heading, available ingredients in database, carousel of images) 
- Includes 4 classes
     - DbConnector (for establishing connection to the database) 
     - Ingredients (a class representing ingredients)
     - IngredientHydrator (for instantiating ingredient objects with available ingredients from the database)  
     - DisplayIngredients (for outputting accordion of ingredients to page) 
- Includes a unit test for the displayIngredients method in DisplayIngredients class

File: app/css/style.css [hunk 1]
```diff
@@ -1,2 +1,92 @@
+* {
```
Comment (mporam) [inline app/css/style.css]: it is common to not commit css files, as they are now auto generated, you can put them in the git ignore

File: app/index.php [hunk 2]
```diff
@@ -2,11 +2,71 @@
...
     <meta charset="UTF-8">
-    <title>Title</title>
-    <link rel="stylesheet" type="text/css" href="css/style.css" >
+    <title>Bread and Ifits</title>
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
+          integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
+    <link rel="stylesheet" type="text/css" href="css/style.css">
+    <link href="https://fonts.googleapis.com/css?family=Raleway&display=swap" rel="stylesheet">
     <script type="text/javascript" src="js/dist/main.js"></script>
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
```
Comment (mporam) [inline app/index.php:16]: 2 bootstrap CSS files?

File: app/index.php [hunk 3]
```diff
@@ -2,11 +2,71 @@
...
+        book</a>
+</nav>
+<div class="nav-transform"></div>
+<div class="container-fluid">
+    <div class="row" id="mainContent">
+        <aside class="col-2 justify-content-center">
+            <form method="post" id="ingredientsForm">
+                <input class="choiceBtn col-2" id="getRecipeBtn" type="submit" value="Get Recipe">
+                <?php
+                require('../vendor/autoload.php');
```
Comment (mporam) [inline app/index.php]: generally i would put the autoloader and use statements at the very top of the file. incase anyone needs to use these things before this point in the page

File: app/index.php [hunk 4]
```diff
@@ -2,11 +2,71 @@
...
+                use BreadAndIfit\DbConnector;
+                use BreadAndIfit\Ingredients\DisplayIngredients;
+                use BreadAndIfit\Ingredients\IngredientHydrator;
 
+                $db = DbConnector::getDatabase();
+
+                $ingredients = IngredientHydrator::getIngredients($db);
+
+                $accordion = DisplayIngredients::displayIngredients($ingredients);
```
Comment (mporam) [inline app/index.php]: just echo this line, u dont need the variable

File: app/scss/style.scss [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+.choiceBtn {
+  background-color: $accentColor1;
+  color: white;
+  padding: 5px 10px;
+  border: none;
+}
+
+main.col-9 {
+  padding-left:0;
```
Comment (mporam) [inline app/scss/style.scss]: missing space after `:`, make sure file is consistent

File: src/Ingredients/DisplayIngredients.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace BreadAndIfit\Ingredients;
+
+class DisplayIngredients
+{
+    /**
+     * This method calls the functions below and its purpose is to be called in the index.php to make
```
Comment (mporam) [inline src/Ingredients/DisplayIngredients.php]: too specific a comment. dont need to start with "This method"... we know its a method. also it doesnt only have to be used in index.php

File: src/Ingredients/DisplayIngredients.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+     * the accordion of ingredients.
+     * @param array $ingredients is the array of objects that came from the database.
+     * @return string is the outputted accordion.
+     */
+    public static function displayIngredients(array $ingredients): string
+    {
+        $ingredientSorted = self::formatData($ingredients);
+
+        $output = self::outputHTML($ingredientSorted);
```
Comment (mporam) [inline src/Ingredients/DisplayIngredients.php]: why not just return this line?

File: src/Ingredients/DisplayIngredients.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     * This method formats the array of data into a multidimensional associative array of ingredients
+     * organised by their category.
+     * @param array $ingredients is the array of objects that came from the database.
+     * @return array is the formatted array.
+     */
+    private static function formatData(array $ingredients): array
+    {
+        $ingredientSorted = [];
+        foreach ($ingredients as $ingredient) {
+            $ingredientSorted[$ingredient->getCategory()][] = $ingredient->getName();
```
Comment (mporam) [inline src/Ingredients/DisplayIngredients.php]: this discards the entire ingredient object except name, while you may only need the name at the moment, this is not advisable. put the whole object into the array then use the name in the htmloutput function

File: src/Ingredients/DisplayIngredients.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    private static function outputHTML(array $ingredientSorted): string
+    {
+        $accordion = '';
+        $cats = '';
+        foreach ($ingredientSorted as $category => $array) {
+            $list = '';
+            $cats .= $category . ' ';
+            foreach ($array as $key => $value) {
+                $list .= '<div><input type="checkbox" value="' . $value . '" name="' . $value . '"><span class="ingredient">'.' '. $value . '</span></input></div>';
```
Comment (mporam) [inline src/Ingredients/DisplayIngredients.php]: opening and closing input tag is invalid HTML. you should use a label, like this:
`<label><input type="checkbox" />Tomato</label>`
that will make the word clickable too which is better UX

File: src/Ingredients/DisplayIngredients.php (hunk 8 above)
Comment (mporam) [reply src/Ingredients/DisplayIngredients.php]: this also negates the purpose of using an object and means the names could be altered before being output accidentally.

File: src/Ingredients/Ingredient.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+namespace BreadAndIfit\Ingredients;
+
+
+class Ingredient
+{
+    protected $id;
+    protected $name;
+    protected $category;
+
+    public function getId(): int
```
Comment (mporam) [inline src/Ingredients/Ingredient.php:16]: doc blocks?

File: src/Ingredients/IngredientHydrator.php [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+     * @param \PDO $db is the database variable established in DBConnector
+     * @return array is the array of objects returned from the DB.
+     */
+    public static function getIngredients(\PDO $db): array
+    {
+        $query = $db->query('SELECT `id`, `name`, `category` FROM ingredients');
+
+        $query->setFetchMode(\PDO::FETCH_CLASS, '\BreadAndIfit\Ingredients\Ingredient');
+
+        $ingredients = $query->fetchAll();
```
Comment (mporam) [inline src/Ingredients/IngredientHydrator.php]: return this line?

**Review by mporam — CHANGES_REQUESTED** [review]

File: app/index.php (hunk 2 above)
Comment (harrycoulton) [reply app/index.php:16]: Removed the first

File: app/scss/style.scss [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+.nav {
+  background-color: $secondaryColor;
+  z-index: 200;
+  height: 50px;
+  width: 100%;
+  position: fixed;
+
+  .nav-link {
+    color: black;
```
Comment (luciavelasco) [inline app/scss/style.scss]: This should be hex.

File: src/Ingredients/DisplayIngredients.php [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+                        <a class="card-header" type="button" data-toggle="collapse" data-target="#' . $category . '"
+                           aria-expanded="false" aria-controls="' . $category . '">' . $category . '</a>
+                        <div class="collapse multi-collapse" id="' . $category . '">
+                            <div class="card-body">
+                               <ul>' . $list . ' </ul>
+                            </div>
+                            </div>';
+        }
+        $accordion .= '</div>';
+        $accordion = '<input type="button" class="choiceBtn" id="expandAll" data-toggle="collapse"
```
Comment (luciavelasco) [inline src/Ingredients/DisplayIngredients.php]: Classes here are camelCase, classes in html are kebab-case, please make them consistent.

File: app/index.php (hunk 2 above)
Comment (luciavelasco) [reply app/index.php:16]: Line 6 and line 11 are still identical

File: app/index.php [hunk 14]
```diff
@@ -25,23 +30,13 @@
...
                 $ingredients = IngredientHydrator::getIngredients($db);
-
-                $accordion = DisplayIngredients::displayIngredients($ingredients);
-
-                echo $accordion;
-
+                echo  DisplayIngredients::displayIngredients($ingredients);
                 ?>
             </form>
+            <input class="choiceBtn col-2" id="getRecipeBtn" type="submit" value="Get Recipe">
```
Comment (mporam) [inline app/index.php]: class name not consistent

File: src/Ingredients/DisplayIngredients.php [hunk 15]
```diff
@@ -59,11 +57,11 @@ private static function outputHTML(array $ingredientSorted): string
                             </div>
                             </div>';
         }
-        $accordion .= '</div>';
-        $accordion = '<input type="button" class="choiceBtn" id="expandAll" data-toggle="collapse"
+            $accordion .= '</div>';
+            $accordion = '<input type="button" class="choiceBtn" id="expandAll" data-toggle="collapse"
```
Comment (mporam) [inline src/Ingredients/DisplayIngredients.php]: class name not consistent

File: src/Ingredients/Ingredient.php [hunk 16]
```diff
@@ -10,16 +10,29 @@ class Ingredient
     protected $name;
     protected $category;
 
+
+    /**
+     * Gets the ID of each ingredient.
+     * @return int the ID from the DB.
+     */
+
```
Comment (mporam) [inline src/Ingredients/Ingredient.php]: no break between doc block and function

File: src/Ingredients/Ingredient.php [hunk 17]
```diff
@@ -10,16 +10,29 @@ class Ingredient
...
     public function getId(): int
     {
         return $this->id;
     }
 
+    /** Gets the name of each ingredient.
```
Comment (mporam) [inline src/Ingredients/Ingredient.php]: descriptions on their own line

File: tests/Ingredients/DisplayIngredientsTest.php [hunk 18]
```diff
@@ -27,7 +27,7 @@ public function testDisplayIngredients_success()
                            aria-expanded="false" aria-controls="Vegetables">Vegetables</a>
                         <div class="collapse multi-collapse" id="Vegetables">
                             <div class="card-body">
-                               <ul><div><input type="checkbox" value="carrot" name="carrot"><span class="ingredient"> carrot</span></input></div> </ul>
+                               <ul><label><input type="checkbox" value="carrot" name="carrot"><span class="ingredient"> carrot</span></input></label> </ul>
```
Comment (mporam) [inline tests/Ingredients/DisplayIngredientsTest.php]: still invalid HTML, inputs cant have closing tags.

**Review by mporam — CHANGES_REQUESTED** [review]

**Review by mporam — CHANGES_REQUESTED** [review]
the front end for showing ingredients is now broken because you have used a label. remember to check the front end before submitting for code review

File: app/scss/style.scss [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+form{
+  text-align: center;
+  padding-bottom: 55px;
+}
+
+.choice-btn {
+  background-color: $accentColor1;
+  color: white;
```
Comment (luciavelasco) [inline app/scss/style.scss:112]: this should be hex, too

File: app/scss/style.scss [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+  }
+}
+
+#carouselExampleControls {
+  padding: 0;
+}
+
+.carousel-inner img {
+  width:100vw;
+  height:100vh;
```
Comment (luciavelasco) [inline app/scss/style.scss:90]: missing space after `:` here, too

**Review by mporam — APPROVED** [review]

**Pull request merged**